server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java

```diff
@@ -0,0 +1,494 @@
+   * @param initProps
+   *          map of k, v pairs of initial properties.
+   */
+  public synchronized static void initSysProps(final ServerContext context,
+      final Map<String,String> initProps) {
+    PropCacheKey sysPropKey = PropCacheKey.forSystem(context.getInstanceID());
+    createInitialProps(context, sysPropKey, initProps);
+  }
+
+  /**
+   * Create initial properties if they do not exist. If the node exists, initialization will be
+   * skipped.
+   *
+   * @param context
+   *          the system context
+   * @param propCacheKey
+   *          a prop id
+   * @param props
+   *          initial properties
+   */
+  public static void createInitialProps(final ServerContext context,
+      final PropCacheKey propCacheKey, Map<String,String> props) {
+
+    try {
+      ZooReaderWriter zrw = context.getZooReaderWriter();
+      if (zrw.exists(propCacheKey.getPath())) {
+        return;
+      }
+      VersionedProperties vProps = new VersionedProperties(props);
+      zrw.putPersistentData(propCacheKey.getPath(), codec.toBytes(vProps),
+          ZooUtil.NodeExistsPolicy.FAIL);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new PropStoreException("Interrupted creating node " + propCacheKey, ex);
+    } catch (Exception ex) {
+      throw new PropStoreException("Failed to create node " + propCacheKey, ex);
+    }
+  }
+
+  public PropStoreMetrics getMetrics() {
+    return cacheMetrics;
+  }
+
+  @Override
+  public void create(PropCacheKey propCacheKey, Map<String,String> props) {
+
+    try {
+      VersionedProperties vProps = new VersionedProperties(props);
+      String path = propCacheKey.getPath();
+      zrw.putPersistentData(path, codec.toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL);
+    } catch (IOException | KeeperException | InterruptedException ex) {
+      throw new PropStoreException("Failed to serialize properties for " + propCacheKey, ex);
+    }
+  }
+
+  /**
+   * get or create properties from the store. If the property node does not exist in ZooKeeper,
+   * legacy properties exist, they will be converted to the new storage form and naming convention.
+   * The legacy properties are deleted once the new node format is written.
+   *
+   * @param propCacheKey
+   *          the prop cache key
+   * @return The versioned properties or null if the properties do not exist for the id.
+   * @throws PropStoreException
+   *           if the updates fails because of an underlying store exception
+   */
+  @Override
+  public @NonNull VersionedProperties get(final PropCacheKey propCacheKey)
+      throws PropStoreException {
+    checkZkConnection(); // if ZK not connected, block, do not just return a cached value.
+    propStoreWatcher.registerListener(propCacheKey, this);
+
+    var props = cache.get(propCacheKey);
+    if (props != null) {
+      return props;
+    }
+
+    return new ConfigTransformer(zrw, codec, propStoreWatcher).transform(propCacheKey);
```

Comment:
Looking at the impl for ConfigTransformer it puts an ehpemeral node in ZK each time its called.  This feels expensive for what seems like it should be a read operation.  Seems like its doing this as part of an upgrade, could this not be done in the Managers upgrade proces as a one time deal instead of every time this is read?

Reply:
889bbf1a15 modified the code so that only the system property node will be created if it does not exist - this allows the tservers to start and wait for assignments. All other property nodes are transformed during the upgrade sequence executed in the master.  

This prevents an issue with the potential of a race condition in the original code to automatically recreating a node (with likely changed, default, no properties) if the property node was deleted.  The namespace / table deletion should also be deleting the conf (parent) node, so it would likely have failed anyway, but there could have been a race if the encoded node deletion triggered a change and it was read / created before the parent node was deleted.

Now, any attempt to read a non-existent node will throw an exception.

The changes in 889bbf1a15 also strengthens the NonNull contract with the exception being thrown if a non-existent node is tried to be read.

Reply:
I looked over the changes when I reviewed, looks good.  The chicken and egg nature of the system props is tricky.